src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineView.xaml.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Windows;
+using System.Windows.Controls;
+using System.Windows.Data;
+using System.Windows.Forms.VisualStyles;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.LanguageServices.Implementation;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.TextManager.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    /// <summary>
+    /// Interaction logic for DocumentOutlineView.xaml
+    /// All operations happen on the UI thread for visual studio
+    /// </summary>
+    internal sealed partial class DocumentOutlineView : UserControl, IVsCodeWindowEvents, IDisposable
+    {
+        private readonly IVsCodeWindow _codeWindow;
+        private readonly DocumentOutlineViewModel _viewModel;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly Dictionary<IVsTextView, ITextView> _trackedTextViews = new();
+        private readonly ComEventSink _codeWindowEventsSink;
+        private readonly ICollectionView _collectionView;
+
+        // Used to suspend all event handlers for caret movement while we navigate the cursor
+        // Should only be written to withing the SymbolTree_MouseDown method which always
+        // is called by WPF on the UI thread.
+        private bool _isNavigating;
+
+        public DocumentOutlineView(
+            DocumentOutlineViewModel viewModel,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            IVsCodeWindow codeWindow)
+        {
+            _codeWindow = codeWindow;
+            _viewModel = viewModel;
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            DataContext = _viewModel;
+            InitializeComponent();
+
+            // "DocumentSymbolItems" is the key name we specified for our CollectionViewSource in the XAML file
+            _collectionView = ((CollectionViewSource)FindResource("DocumentSymbolItems")).View;
+
+            // We don't think the shell is initialized lazily, so we'll Debug.Fail(), but if it was we'd still
+            // see the view created later so this will still function.
+            if (ErrorHandler.Failed(codeWindow.GetPrimaryView(out var primaryTextView)))
```

Comment:
We need to refactor this out since we also should share this with navbar -- I had mentioned this in an earlier PR and clearly forgot to do the cleanup. I'm writing this here mostly so hopefully I won't forget this a second time. :smile:

Reply:
https://github.com/dotnet/roslyn/issues/66448